internal/app/upstream/client.go

```diff
@@ -55,11 +74,177 @@ type Response struct {
 //
 // The method does not block until the underlying connection is up.
 // Returns immediately and connecting the server happens in background
-// TODO: pass retry/timeout configurations
-func NewClient(ctx context.Context, url string) (Client, error) {
-	return &client{}, nil
+// TODO: pass retry configurations
+func NewClient(ctx context.Context, url string, callOptions CallOptions) (Client, error) {
+	conn, err := grpc.Dial(url, grpc.WithInsecure())
+	if err != nil {
+		return nil, err
+	}
+
+	ldsClient := v2.NewListenerDiscoveryServiceClient(conn)
+	rdsClient := v2.NewRouteDiscoveryServiceClient(conn)
+	edsClient := v2.NewEndpointDiscoveryServiceClient(conn)
+	cdsClient := v2.NewClusterDiscoveryServiceClient(conn)
+
+	go shutDown(ctx, conn)
+
+	return &client{
+		ldsClient:   ldsClient,
+		rdsClient:   rdsClient,
+		edsClient:   edsClient,
+		cdsClient:   cdsClient,
+		callOptions: callOptions,
+	}, nil
+}
+
+// NewMockClient creates a mock implementation for testing
+func NewMockClient(
```

Comment:
What are the best practices for exposing Mock functions?

Reply:
There are multiple ways to do it. What you are doing is fine. The other way I have seen is moving this method to a fake.go file that has access to all the private fields/methods.